packages/fuels-programs/src/logs.rs

```diff
@@ -34,26 +69,24 @@ impl LogDecoder {
             .iter()
             .extract_log_id_and_data()
             .filter_map(|(log_id, data)| {
-                self.type_lookup
+                self.log_formatters
                     .get(&log_id)
-                    .map(|param_type| param_type.decode_log(&data))
+                    .map(|log_formatter| log_formatter.format(&data))
             })
             .collect()
     }
 
     /// Get decoded logs with specific type from the given receipts.
     /// Note that this method returns the actual type and not a `String` representation.
-    pub fn get_logs_with_type<T: Tokenizable + Parameterize>(
+    pub fn get_logs_with_type<T: Tokenizable + Parameterize + 'static>(
         &self,
         receipts: &[Receipt],
     ) -> Result<Vec<T>> {
-        let target_param_type = T::param_type();
-
         let target_ids: HashSet<LogId> = self
-            .type_lookup
+            .log_formatters
             .iter()
-            .filter_map(|(log_id, param_type)| {
-                (*param_type == target_param_type).then_some(log_id.clone())
+            .filter_map(|(log_id, log_formatter)| {
+                log_formatter.can_handle_type::<T>().then(|| log_id.clone())
```

Comment:
It's not as if we fetch type information from the formatter but rather ask if it is capable of decoding the given type. 

Since the idea of loggers being able to decode only some type (here only one type per logger instance), being able to ask them if they can decode a type seems reasonably within their concerns.

From the outside, it just looks like an innocent question that is easy to answer. While the implementation goes into the TypeId business.